docs/reference/ci-cd.md

```diff
@@ -1,42 +1,81 @@
 
+(cicd/hub)=
 ## Automatic hub deployment
 
+% TODO: Add a link to the blog post when it is published
+
 The best place to learn about the latest state of our *automatic* hub deployment
 is to look at [the `deploy-hubs.yaml` GitHub Actions workflow file](https://github.com/2i2c-org/infrastructure/tree/HEAD/.github/workflows/deploy-hubs.yaml).
-That workflow defines the automatic hub deployment for many of our major clusters.
+This workflow file depends on a locally defined action that [sets up access to a given cluster](https://github.com/2i2c-org/infrastructure/blob/master/.github/actions/setup-deploy/action.yaml) and itself contains four main jobs, detailed below.
+
+(cicd/hub/job1)=
+### 1. `generate-jobs`: Generate Helm upgrade jobs
+
+The first job takes a list of files that have been added/modified as part of a Pull Request and pipes them into the [`generate-helm-upgrade-jobs` sub-command](https://github.com/2i2c-org/infrastructure/blob/master/deployer/helm_upgrade_decision.py) of the [deployer module](https://github.com/2i2c-org/infrastructure/tree/master/deployer).
+This sub-command uses a set of functions to calculate which hubs on which clusters require a helm upgrade, alongside whether the support chart and staging hub on that cluster should also be upgraded.
+If any production hubs require an upgrade, the upgrade of the staging hub is a requirement.
+
+This job provides the following outputs:
+
+- Two JSON objects that can be read by later GitHub Actions jobs to define matrix jobs.
+  These JSON objects detail: which clusters require their support chart and/or staging hub to be upgraded, and which production hubs require an upgrade.
+- The above JSON objects are also rendered as human-readable tables using [`rich`](https://github.com/Textualize/rich).
 
-Currently, our CI/CD system is triggered on pushes to the default branch when any of the
-following paths are modified:
+#### Some special cased filepaths
 
+While the aim of this workflow is to only upgrade the pieces of the infrastructure that require it with every change, some changes do require us to redeploy everything.
+
+- If a cluster's `cluster.yaml` file has been modified, we upgrade the support chart and **all** hubs on **that** cluster. This is because we cannot tell what has been changed without inspecting the diff of the file.
+- If either the `basehub` or `daskhub` Helm charts have additions/modifications in their paths, we redeploy **all** hubs across **all** clusters.
+- If the support Helm chart has additions/modifications in its path, we redeploy the support chart on **all** clusters.
+- If the deployer module has additions/modifications in its path, then we redeploy **all** hubs on **all** clusters.
+
+```{note}
+Right now, we redeploy everything was the deployer changes since the deployer undertakes some tasks that generates config related to authentication.
+This may change in the future as we move towards the deployer becoming a separable, stand-alone package.
```

Comment:
*I think*?
```suggestion
Right now, we redeploy everything when the deployer changes since the deployer undertakes some tasks that generates config related to authentication.
This may change in the future as we move towards the deployer becoming a separable, stand-alone package.
```